### PR TITLE
Transform Messages Capability

### PR DESCRIPTION
**Pull request by WaelKarkoub**
## Why are these changes needed?

When I think of transforming chat messages (truncating, compressing, summarizing, etc.), it sounded to me like it's a transformation matrix problem in robotics. For example:

$\textbf{P}_2 = \textbf{R}_x \textbf{R}_y \textbf{R}_z \textbf{P}_1$ 

Where $\textbf{R}$ is the rotation transformation matrix relative to an axis, and $\textbf{P}$ is the position vector.

So I thought, why not approach transforming chat messages the same way?

$\textbf{M}_2 = \textbf{R}_a \textbf{R}_b \textbf{R}_c \textbf{M}_1$

 For example:
- $\textbf{R}_a$ can be max number of messages,
- $\textbf{R}_b$ can be token limit,
- $\textbf{R}_c$ LLM Lingua compression

So I refactored `TransformChatHistory` in `context_handling.py` to get the same behavior described above, cc @gagb. So the ability is now called `TransformMessages`, and it accepts a list of the transformations you would like to apply to the messages. `MessageTransform` defines the interface of how the transform, $\textbf{R}$, should look like. I also implemented a few transforms as early examples (`MaxMessagesTransform`, `TruncateMessageTransform`)

### Why is this useful?
There are some text-based use cases like I mentioned above that can be very helpful in managing the number of tokens used, but they can also be useful for other modalities. Imagine a certain LLM can only accept a certain image size, but the chat history has all kinds of image sizes and resolutions. All you have to do is create a transform (e.g. `ResizeImageTransfrom`) and pass it to the ability.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by gagb** [comment]
This is nice! I'll check @dkirsche might also be interested

**Comment by gagb** [comment]
@WaelKarkoub I am running late on this PR-- I will have time after 14th to review this but I will do this. So sorry :)

**Comment by gagb** [comment]
@WaelKarkoub LFG!!! This align so much with what I am used to with other libs. Feedback:
1. Add NB demonstrating your code in use
2. Add test cases

Wdyt?

**Comment by WaelKarkoub** [comment]
@gagb notebook added :heavy_check_mark: tests added :heavy_check_mark:

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (gagb) [inline autogen/agentchat/contrib/capabilities/transform_messages.py:116]: Improve doc to clarify the strategy it uses to limit.

File: notebook/agentchat_transform_messages.ipynb
Comment (gagb) [inline notebook/agentchat_transform_messages.ipynb:1]: Fantastic job reusing existing example and showing compatibility.

File: test/agentchat/contrib/capabilities/test_transform_messages.py
Comment (gagb) [inline test/agentchat/contrib/capabilities/test_transform_messages.py:104]: Why is this here?

File: test/agentchat/contrib/capabilities/test_transform_messages.py
Comment (gagb) [inline test/agentchat/contrib/capabilities/test_transform_messages.py:155]: Ensure tests following format in other files. Eg `test_context_handling`. Main function is missing.

**Review by gagb — COMMENTED** [review]
Ty for changes. See new comments.

File: test/agentchat/contrib/capabilities/test_transform_messages.py
Comment (WaelKarkoub) [reply test/agentchat/contrib/capabilities/test_transform_messages.py]: Removed!

**Comment by WaelKarkoub** [comment]
@gagb Thanks for the review! I updated the docstrs, let me know what you think of them

**Comment by gagb** [comment]
> @gagb Thanks for the review! I updated the docstrs, let me know what you think of them

Will review today.

**Comment by gagb** [comment]
@WaelKarkoub see suggested changes here: https://github.com/microsoft/autogen/pull/2125

**Review by gagb — COMMENTED** [review]
@WaelKarkoub see suggested changes here: https://github.com/microsoft/autogen/pull/2125.

Please lmk if these make sense, and I can approve it. After this merge perhaps we can add transformation for tools and retire the previous context handling module, since this one is more general.

**Comment by BeibinLi** [comment]
This is a great PR. Thanks so much!

@sonichi  @ekzhu  @qingyun-wu  @jackgerrits  We had conversation about creating our own "message" format. Do you think we should proceed ASAP? This PR could be a good example converting TextMessage to ImageMessage or AudioMessage, etc.


For instance, we can have
```python
class Message(Protocol):
     ...

class TextMessage:
    ...

class ImageMessage:
    ...
```

**Comment by WaelKarkoub** [comment]
@BeibinLi awesome idea! Yes, let us pause this PR till these new message formats go in, should make it easier to refactor for everyone. Let me know if I can help

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (BeibinLi) [inline autogen/agentchat/contrib/capabilities/transform_messages.py:113]: This is a good example, which can help users and engineers how to use "MessageTransform".

FYI, it is related to the [Compressible Agent](https://github.com/microsoft/autogen/blob/main/autogen/agentchat/contrib/compressible_agent.py).

Maybe, link the compressible agent to the docstring here for reference. Such as 
```markdown
For more sophisticated memory management (utilizing vector databases), please refer to the [Compressible Agent](https://github.com/microsoft/autogen/blob/main/autogen/agentchat/contrib/compressible_agent.py).
```

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (BeibinLi) [inline autogen/agentchat/contrib/capabilities/transform_messages.py:145]: Sometimes, it is better to remove the messages in the middle rather than the head, because often the first 1 or 2 messages contains the highest-level of task.

Should we add a parameter in the __init__ function, `truncate_strategy: str = "head"`, where it could be "head", "middle", etc.

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (BeibinLi) [inline autogen/agentchat/contrib/capabilities/transform_messages.py:267]: Since we have `self._model`, should we also make sure the max_tokens is less than the model's capacity? 

```python
allowed_tokens = token_count_utils.get_max_token_limit(self._model)
if max_tokens > allowed_tokens:
    logger.warning(f"Max token was set to {max_tokens}, but {self._model} can only accept {allowed_tokens} tokens. Capping it to {allowed_tokens}.")
    max_tokens = allowed_tokens
```

**Review by BeibinLi — COMMENTED** [review]
Thanks for this awesome PR! I made a few Lint comments. The most important thing is about function name. All other comments are auxiliary.

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (gagb) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:18]: Or just `transform`? I kind of liked the explicit syntax.

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (gagb) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:113]: @BeibinLi for context it generalizes the existing `context_handling.py` capability

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:18]: I also like the explicit syntax. The issue with setting up the call method is that it won't be clear to the user what that class might be doing at first glance, adding cognitive load.

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:113]: Yes, the `MessageHistoryLimiter` is extracted from `context_handling.py`, which should be deprecated if this goes in? 

However, I do have a follow-up PR to add a transform to replace the Compressible Agent, by either implementing the same logic or use [LLMLingua](https://github.com/microsoft/LLMLingua) since it's also a Microsoft project

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:145]: Interesting idea, I didn't think about this. What do you think about the idea of having a single transform per strategy? I'm also thinking about doing this to `MessageTokenLimiter`, splitting it up into `HistoryTokenLimiter` and `MessageTokenLimiter` (still didn't finalize the names).

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:267]: Very cool! Didn't know this existed. I'll add it as a validator.

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/transform_messages.py:145]: Also for the implementation, maybe the argument could `skip: Optional[List[int]] = None`, and if the user wants to skip the first two messages, he can do `skip=[0,1]`

File: autogen/agentchat/contrib/capabilities/transforms.py
Comment (rickyloynd-microsoft) [inline autogen/agentchat/contrib/capabilities/transforms.py:88]: The current docstring could be a bit more specific on the cutoff boundary. "If any message raises the accumulated token beyond this limit, that message as well as any remaining messages are discarded."

File: autogen/agentchat/contrib/capabilities/transforms.py
Comment (rickyloynd-microsoft) [inline autogen/agentchat/contrib/capabilities/transforms.py:89]: "prepending messages"

File: autogen/agentchat/contrib/capabilities/transforms.py
Comment (rickyloynd-microsoft) [inline autogen/agentchat/contrib/capabilities/transforms.py:137]: "prepend the message"

File: autogen/agentchat/contrib/capabilities/transform_messages.py
Comment (gagb) [reply autogen/agentchat/contrib/capabilities/transform_messages.py]: Yes we should deprecate that module once your solution gets in!

**Review by rickyloynd-microsoft — APPROVED** [review]
Great work!

**Comment by WaelKarkoub** [comment]
> This is a great PR. Thanks so much!
> 
> @sonichi @ekzhu @qingyun-wu @jackgerrits We had conversation about creating our own "message" format. Do you think we should proceed ASAP? This PR could be a good example converting TextMessage to ImageMessage or AudioMessage, etc.
> 
> For instance, we can have
> 
> ```python
> class Message(Protocol):
>      ...
> 
> class TextMessage:
>     ...
> 
> class ImageMessage:
>     ...
> ```



> @BeibinLi awesome idea! Yes, let us pause this PR till these new message formats go in, should make it easier to refactor for everyone. Let me know if I can help

@gagb @rickyloynd-microsoft I'll be pausing development till the proposed new message format goes in, should I send it back to draft?

**Comment by gagb** [comment]
> > This is a great PR. Thanks so much!
> > @sonichi @ekzhu @qingyun-wu @jackgerrits We had conversation about creating our own "message" format. Do you think we should proceed ASAP? This PR could be a good example converting TextMessage to ImageMessage or AudioMessage, etc.
> > For instance, we can have
> > ```python
> > class Message(Protocol):
> >      ...
> > 
> > class TextMessage:
> >     ...
> > 
> > class ImageMessage:
> >     ...
> > ```
> 
> > @BeibinLi awesome idea! Yes, let us pause this PR till these new message formats go in, should make it easier to refactor for everyone. Let me know if I can help
> 
> @gagb @rickyloynd-microsoft I'll be pausing development till the proposed new message format goes in, should I send it back to draft?

I need @sonichi help to make that decision. I am personally okay with it for following reasons:
- nobody is blocked
- if someone starts using your contribution and we introduced breaking changes, that'd create overhead to communicate
On the other hand, if the message protocol blocks this PR for 2 months, that won't be ideal?

**Comment by ekzhu** [comment]
I feel the same as @rickyloynd-microsoft and @gagb. A message protocol update can be a big update. It won't be ideal for this PR to be blocked by that.

**Comment by gagb** [comment]
@WaelKarkoub then it seems like other maintainers agree with proceeding with this PR. wdyt?

**Comment by WaelKarkoub** [comment]
@gagb sounds good!

@BeibinLi I just have a couple of unresolved comments and should be good to merge

File: notebook/agentchat_transform_messages.ipynb
Comment (sonichi) [inline notebook/agentchat_transform_messages.ipynb:1]: Shall we make this notebook renderable on the website?
cc @jackgerrits

File: autogen/agentchat/contrib/capabilities/context_handling.py
Comment (sonichi) [inline autogen/agentchat/contrib/capabilities/context_handling.py:11]: Shall we add a note to the corresponding notebook?

**Review by BeibinLi — APPROVED** [review]
LGTM.

**Comment by jackgerrits** [comment]
I think we should proceed with this PR. I am working on at least solidifying the existing assumptions we have in place around message formats in a PR I hope to open soon. Once assumptions are solidified it will be clearer what the extensions we have in place are and how to type them

File: autogen/agentchat/contrib/capabilities/context_handling.py
Comment (WaelKarkoub) [reply autogen/agentchat/contrib/capabilities/context_handling.py:11]: Just to clarify, a note to deprecate the notebook or a note to tell the user about the new transform messages notebook?

File: notebook/agentchat_transform_messages.ipynb
Comment (WaelKarkoub) [reply notebook/agentchat_transform_messages.ipynb:1]: is it as simple as adding:
```json
"front_matter": {
   "description": "Preprocessing chat history with `TransformMessages`",
   "tags": [
    "long context handling",
    "capability"
   ]
  }
```

File: autogen/agentchat/contrib/capabilities/context_handling.py
Comment (sonichi) [reply autogen/agentchat/contrib/capabilities/context_handling.py:11]: I meant that there is a notebook using the old capability. We can add a note in that notebook about this and a link to the new notebook too.

File: notebook/agentchat_transform_messages.ipynb
Comment (sonichi) [reply notebook/agentchat_transform_messages.ipynb:1]: @jackgerrits could you answer this question? Also consider making the instructions easier to find.

**Comment by sonichi** [comment]
@gagb could you confirm that the old way of handling context will be deprecated? Shall we remove the test and eventually remove the code and examples too?
@jackgerrits there is one comment that needs your input.

**Comment by gagb** [comment]
> @gagb could you confirm that the old way of handling context will be deprecated? Shall we remove the test and eventually remove the code and examples too? @jackgerrits there is one comment that needs your input.

Yes old way should be deprecated. The whole point of this solution was to generalize.

File: notebook/agentchat_transform_messages.ipynb
Comment (sonichi) [reply notebook/agentchat_transform_messages.ipynb:1]: #2191

**Comment by WaelKarkoub** [comment]
@sonichi @jackgerrits updated the PR, notebook should be renderable and formatted to follow the guidelines

**Review by sonichi — APPROVED** [review]
@gagb could you add the deprecation plan to the roadmap?

**Pull request merged**
